From 2a35cec5bc056040c3beb733f65751311fa3d1c1 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Tue, 15 Jan 2013 10:06:08 -0700
Subject: [PATCH] Add Change specific tokens to ChangeDahsboards

Add the following tokens to be used for ChangeDashboards:

 ${branch}
 ${change.id}
 ${change.key}
 ${project}
 ${ref}
 ${subject}
 ${topic}

Change-Id: Id38c35de2535d45b3d0488cf0fb84bcd582183d1
---
 Documentation/user-dashboards.txt                  | 12 +++++++
 .../rpc/changedetail/ChangeDetailFactory.java      | 15 +++++++-
 .../gerrit/server/project/DashboardResource.java   | 34 ++++++++++++++++++
 .../server/project/DashboardsCollection.java       | 40 ++++++++++++++++++----
 .../google/gerrit/server/project/GetDashboard.java |  8 +++--
 .../gerrit/server/project/ListDashboards.java      |  1 +
 6 files changed, 100 insertions(+), 10 deletions(-)

diff --git a/Documentation/user-dashboards.txt b/Documentation/user-dashboards.txt
index 2910491..55dac78 100644
--- a/Documentation/user-dashboards.txt
+++ b/Documentation/user-dashboards.txt
@@ -171,6 +171,18 @@ the child projects.
   local-change = refs/meta/dashboards/main:local_change
 ----
 
+Change dashboard queries may contain the following additional tokens
+which will be replaced with change specific data:
+
+ ${branch}
+ ${change.id}
+ ${change.key}
+ ${project}
+ ${ref}
+ ${subject}
+ ${topic}
+
+
 GERRIT
 ------
 Part of link:index.html[Gerrit Code Review]
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
index 818df5b..5678a44 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
@@ -58,6 +58,8 @@ import com.google.inject.Inject;
 import com.google.inject.Provider;
 import com.google.inject.assistedinject.Assisted;
 
+import static com.google.common.base.Objects.firstNonNull;
+
 import org.eclipse.jgit.errors.ConfigInvalidException;
 import org.eclipse.jgit.errors.RepositoryNotFoundException;
 import org.eclipse.jgit.lib.Config;
@@ -202,9 +204,20 @@ public class ChangeDetailFactory extends Handler<ChangeDetail> {
   }
 
   private void loadDashboard() {
+    DashboardResource.HashMapTokenizer tokenizer =
+        new DashboardResource.HashMapTokenizer();
+      Change change = detail.getChange();
+    tokenizer.put("branch", change.getDest().getShortName());
+    tokenizer.put("change.id", Integer.toString(change.getId().get()));
+    tokenizer.put("change.key", change.getKey().get());
+    tokenizer.put("project",change.getDest().getParentKey().toString());
+    tokenizer.put("ref", change.getDest().get());
+    tokenizer.put("subject", change.getSubject());
+    tokenizer.put("topic", firstNonNull(change.getTopic(), ""));
+
     GetDashboard getDashboard = dashboardProvider.get();
     DashboardResource resource = DashboardResource.projectTyped(
-        control.getProjectControl(), Project.DashboardType.CHANGE);
+        control.getProjectControl(), Project.DashboardType.CHANGE, tokenizer);
     try {
       DashboardInfo info = getDashboard.apply(resource);
       detail.setDashboardTitle(info.title);
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java
index 4672ff5..6bb000d 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardResource.java
@@ -21,20 +21,40 @@ import com.google.inject.TypeLiteral;
 
 import org.eclipse.jgit.lib.Config;
 
+import java.util.HashMap;
+
 public class DashboardResource implements RestResource {
   public static final TypeLiteral<RestView<DashboardResource>> DASHBOARD_KIND =
       new TypeLiteral<RestView<DashboardResource>>() {};
 
+  public interface Tokenizer {
+    public String tokenize(String value);
+  }
+
+  public static class HashMapTokenizer extends HashMap<String,String>
+      implements Tokenizer {
+    @Override
+    public String tokenize(String value) {
+      return get(value);
+    }
+  }
+
   public static DashboardResource projectTyped(ProjectControl ctl,
       Project.DashboardType type) {
     return new DashboardResource(ctl, null, null, null, type);
   }
 
+  public static DashboardResource projectTyped(ProjectControl ctl,
+      Project.DashboardType type, Tokenizer tokenizer) {
+    return new DashboardResource(ctl, null, null, null, type, tokenizer);
+  }
+
   private final ProjectControl control;
   private final String refName;
   private final String pathName;
   private final Config config;
   private Project.DashboardType type;
+  private Tokenizer tokenizer;
 
   DashboardResource(ProjectControl control,
       String refName,
@@ -48,11 +68,21 @@ public class DashboardResource implements RestResource {
       String pathName,
       Config config,
       Project.DashboardType type) {
+    this(control, refName, pathName, config, null, null);
+  }
+
+  DashboardResource(ProjectControl control,
+      String refName,
+      String pathName,
+      Config config,
+      Project.DashboardType type,
+      Tokenizer tokenizer) {
     this.control = control;
     this.refName = refName;
     this.pathName = pathName;
     this.config = config;
     this.type = type;
+    this.tokenizer = tokenizer;
   }
 
   public ProjectControl getControl() {
@@ -74,4 +104,8 @@ public class DashboardResource implements RestResource {
   public Project.DashboardType getType() {
     return type;
   }
+
+  public Tokenizer getTokenizer() {
+    return tokenizer;
+  }
 }
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java
index db69c7e..7c49fcf 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/DashboardsCollection.java
@@ -50,10 +50,14 @@ import java.util.List;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public class DashboardsCollection implements
     ChildCollection<ProjectResource, DashboardResource>,
     AcceptsCreate<ProjectResource>{
+  private static final Pattern TOKEN_RE = Pattern.compile("[$][{]([^}]*)[}]");
+
   private final GitRepositoryManager gitManager;
   private final DynamicMap<RestView<DashboardResource>> views;
   private final Provider<ListDashboards> list;
@@ -89,10 +93,16 @@ public class DashboardsCollection implements
   @Override
   public DashboardResource parse(ProjectResource parent, String id)
       throws ResourceNotFoundException, IOException, ConfigInvalidException {
+    return parse(parent, id, null);
+  }
+
+  public DashboardResource parse(ProjectResource parent, String id,
+      DashboardResource.Tokenizer tokenizer)
+      throws ResourceNotFoundException, IOException, ConfigInvalidException {
     ProjectControl myCtl = parent.getControl();
     Project.DashboardType type = Project.DashboardType.fromId(id);
     if (type != null) {
-      return DashboardResource.projectTyped(myCtl, type);
+      return DashboardResource.projectTyped(myCtl, type, tokenizer);
     }
 
     List<String> parts = Lists.newArrayList(
@@ -107,7 +117,7 @@ public class DashboardsCollection implements
     Set<Project.NameKey> seen = Sets.newHashSet(ctl.getProject().getNameKey());
     for (;;) {
       try {
-        return parse(ctl, ref, path, myCtl);
+        return parse(ctl, ref, path, myCtl, tokenizer);
       } catch (AmbiguousObjectException e) {
         throw new ResourceNotFoundException(id);
       } catch (IncorrectObjectTypeException e) {
@@ -124,7 +134,7 @@ public class DashboardsCollection implements
   }
 
   private DashboardResource parse(ProjectControl ctl, String ref, String path,
-      ProjectControl myCtl)
+      ProjectControl myCtl, DashboardResource.Tokenizer tokenizer)
       throws ResourceNotFoundException, IOException, AmbiguousObjectException,
           IncorrectObjectTypeException, ConfigInvalidException {
     String id = ref + ":" + path;
@@ -148,7 +158,7 @@ public class DashboardsCollection implements
         throw new ResourceNotFoundException(id);
       }
       BlobBasedConfig cfg = new BlobBasedConfig(null, git, objId);
-      return new DashboardResource(myCtl, ref, path, cfg);
+      return new DashboardResource(myCtl, ref, path, cfg, null, tokenizer);
     } finally {
       git.close();
     }
@@ -160,8 +170,8 @@ public class DashboardsCollection implements
   }
 
   static DashboardInfo parse(Project definingProject, String refName,
-      String path, Config config, String project,
-      Set<Project.DashboardType> setTypes)
+      String path, Config config, DashboardResource.Tokenizer tokenizer,
+      String project, Set<Project.DashboardType> setTypes)
       throws UnsupportedEncodingException {
     DashboardInfo info = new DashboardInfo(refName, path);
     info.project = project;
@@ -188,6 +198,7 @@ public class DashboardsCollection implements
       Section s = new Section();
       s.name = name;
       s.query = config.getString("section", name, "query");
+      s.query = tokenize(s.query, tokenizer);
       u.put(s.name, replace(project, s.query));
       info.sections.add(s);
     }
@@ -197,6 +208,23 @@ public class DashboardsCollection implements
     return info;
   }
 
+  private static String tokenize(String query, DashboardResource.Tokenizer tokenizer) {
+    if (tokenizer == null) {
+      return query;
+    }
+
+    Matcher m = TOKEN_RE.matcher(query);
+    StringBuffer newQuery = new StringBuffer();
+    while (m.find()) {
+      String replacement = tokenizer.tokenize(m.group(1));
+      if (replacement != null) {
+        m.appendReplacement(newQuery, replacement);
+      }
+    }
+    m.appendTail(newQuery);
+    return newQuery.toString();
+  }
+
   private static String replace(String project, String query) {
     return query.replace("${project}", project);
   }
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java
index c3461fc..6dbe356 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/GetDashboard.java
@@ -54,7 +54,7 @@ public class GetDashboard implements RestReadView<DashboardResource> {
     String project = resource.getControl().getProject().getName();
     if (type != null) {
       // The type is not resolved to a definition yet.
-      resource = typeOf(resource.getControl(), type);
+      resource = typeOf(resource.getControl(), resource.getTokenizer(), type);
     }
 
     return DashboardsCollection.parse(
@@ -62,11 +62,13 @@ public class GetDashboard implements RestReadView<DashboardResource> {
         resource.getRefName().substring(REFS_DASHBOARDS.length()),
         resource.getPathName(),
         resource.getConfig(),
+        resource.getTokenizer(),
         project,
         Collections.singleton(type));
   }
 
-  private DashboardResource typeOf(ProjectControl ctl, Project.DashboardType type)
+  private DashboardResource typeOf(ProjectControl ctl, DashboardResource.Tokenizer
+      tokenizer, Project.DashboardType type)
       throws ResourceNotFoundException, IOException, ConfigInvalidException {
     String id = ctl.getProject().getLocalDashboard(type);
     if (Strings.isNullOrEmpty(id)) {
@@ -75,7 +77,7 @@ public class GetDashboard implements RestReadView<DashboardResource> {
     if (Project.DashboardType.fromId(id) != null) {
       throw new ResourceNotFoundException();
     } else if (!Strings.isNullOrEmpty(id)) {
-      return dashboards.parse(new ProjectResource(ctl), id);
+      return dashboards.parse(new ProjectResource(ctl), id, tokenizer);
     } else if (!inherited) {
       throw new ResourceNotFoundException();
     }
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java
index 3c8d74c..d6c4d03 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/ListDashboards.java
@@ -130,6 +130,7 @@ class ListDashboards implements RestReadView<ProjectResource> {
                 ref.getName().substring(REFS_DASHBOARDS.length()),
                 tw.getPathString(),
                 new BlobBasedConfig(null, git, tw.getObjectId(0)),
+                null,
                 project,
                 setTypes));
           } catch (ConfigInvalidException e) {
-- 
1.8.1.rc2

